src/quantum/util/impl/quantum_sequencer_impl.h

```diff
@@ -252,116 +243,131 @@ template <class SequenceKey, class Hash, class KeyEqual, class Allocator>
 template <class FUNC, class ... ARGS>
 int
 Sequencer<SequenceKey, Hash, KeyEqual, Allocator>::waitForTwoDependents(
-    CoroContextPtr<int> ctx,
-    ICoroContextBasePtr dependent,
-    ICoroContextBasePtr universalContext,
-    SequenceKeyStatisticsWriter& stats,
-    void* opaque,
-    const typename Sequencer<SequenceKey, Hash, KeyEqual, Allocator>::ExceptionCallback& exceptionCallback,
-    FUNC&& func,
-    ARGS&&... args)
+        CoroContextPtr<int> ctx,
+        void* opaque,
+        Sequencer& sequencer,
+        SequenceKeyData&& dependent,
+        SequenceKeyData&& universalDependent,
+        FUNC&& func,
+        ARGS&&... args)
 {
     // wait until all the dependents are done
-    if (dependent)
+    if (dependent._context)
     {
-        dependent->wait(ctx);
+        dependent._context->wait(ctx);
     }
-    if (universalContext)
+    if (universalDependent._context)
     {
-        universalContext->wait(ctx);
+        universalDependent._context->wait(ctx);
     }
-
-    // statistics update action to be called after the func, even if func throws
-    auto updateStatsAction = [&stats]()
-    {
-        // update the dependent's stats only, because the task is associated with the non-universal key
-        stats.decrementPendingTaskCount();
-    };
-    callPosted(ctx, 
-               updateStatsAction, 
-               opaque,
-               exceptionCallback, 
-               std::forward<FUNC>(func), 
-               std::forward<ARGS>(args)...);
+    // update task stats
+    dependent._stats->decrementPendingTaskCount();
+    sequencer._taskStats->decrementPendingTaskCount();
+    callPosted(ctx, opaque, sequencer, std::forward<FUNC>(func), std::forward<ARGS>(args)...);
     return 0;
 }
 
 template <class SequenceKey, class Hash, class KeyEqual, class Allocator>
 template <class FUNC, class ... ARGS>
 int
 Sequencer<SequenceKey, Hash, KeyEqual, Allocator>::waitForDependents(
-    CoroContextPtr<int> ctx,
-    std::vector<std::pair<ICoroContextBasePtr, SequenceKeyStatisticsWriter*>>&& dependents,
-    void* opaque,
-    const typename Sequencer<SequenceKey, Hash, KeyEqual, Allocator>::ExceptionCallback& exceptionCallback,
-    FUNC&& func,
-    ARGS&&... args)
+        CoroContextPtr<int> ctx,
+        void* opaque,
+        Sequencer& sequencer,
+        std::vector<SequenceKeyData>&& dependents,
+        SequenceKeyData&& universalDependent,
+        FUNC&& func,
+        ARGS&&... args)
 {
     // wait until all the dependents are done
     for (const auto& dependent : dependents)
     {
-        if (dependent.first)
+        if (dependent._context)
         {
-            dependent.first->wait(ctx);
+            dependent._context->wait(ctx);
         }
+        dependent._stats->decrementPendingTaskCount();
```

Comment:
Notice that if a task is waiting for 2 dependents, and the second dependent is busy, then the pending task count for the 1st dependent may be updated long before callPosted is called. How about running a separate loop to update the dependent stats after the universal dependent waiting is done?

Reply:
Yes I can do that. It will also be more in line with `waitForTwoDependants` where the count is decremented _after_ the universal dependent wait is complete.